README
Distribution Files for UMB Scheme $Revision: 3.2 $

README			This list of files.
scheme.texinfo		UMB Scheme Release Notes (supplied as hardcopy).
Makefile		For compiling UMB Scheme.
prelude.scheme		The Scheme prelude -- primitives loaded at startup.
scheme.1		UMB Scheme manual page (in man(1) format).


Source files:

portable.h		Portability considerations.

steering.c		UMB Scheme steering -- including main()
steering.h

debug.c              	Debugger steering and debug primitives.
debug.h

architecture.c		Registers, stacks, heap, name handling.
architecture.h

object.c		Scheme objects.
object.h

io.c			Read, display, print, file handling.
io.h

compiler.c		Compiler: Expression -> Graph.
compiler.h

eval.c			Eval().
eval.h

primitive.c		Primitives implemented in C (other than numbers).
primitive.h

number.c		Number primitives.
number.h	

bignum.c		Bignum (integers of arb magnitude) support.
bignum.h

complex.c		Complex number support.
complex.h

fixnum.c		Fixnum (smaller integers) support.
fixnum.h

real.c			Real number (C double) support.
real.h

rational.c		NB: rationals not yet implemented (skeletons).
rational.h

SLIB-for-umb-scheme.init
slib/			A copy of the Portable Scheme Library (see copyrights)

See the UMB Scheme Release Notes (scheme.texinfo) for installation instructions.





slib/ChangeLog
Mon Jun 12 16:20:54 1995  Aubrey Jaffer  (jaffer@jacal)

	* synclo.scm (internal-syntactic-environment
	top-level-syntactic-environment): replaced call to alist-copy.

	* require.scm (*catalog*): 'schelog, 'primes, and 'batch added.
	'prime renamed to 'factor.

	From: mhc@edsdrd.eds.com (Michael H Coffin)
	* primes.scm (primes probably-prime?): added.  prime.scm renamed
	to factor.scm.

Fri Mar 24 23:35:25 1995  Matthew McDonald <mafm@cs.uwa.edu.au>

	* struct.scm (define-record): added field-setters.

Sun Jun 11 23:36:55 1995  Aubrey Jaffer  (jaffer@jacal)

	* batch.scm: added

	* Makefile (schelogfiles): SLIB schelog distribution created.

Mon Apr 17 15:57:32 1995  Aubrey Jaffer  (jaffer@jacal)

	* comlist.scm (coerce type-of): added.

	* debug.scm (debug:qp): with *qp-width* of 0 just `write's.

	* paramlst.scm (getopt->parameter-list): Now accepts long-named
	options.  Now COERCEs according to types.

Sat Apr 15 23:15:26 1995  Aubrey Jaffer  (jaffer@jacal)

	* require.scm (require:feature->path): Returns #f instead of
	string if feature not in *catalog* or *modules*.

Sun Mar 19 22:26:52 1995  Aubrey Jaffer  (jaffer@jacal)

	* getopt.scm (getopt-- argc argv optstring): added wrapper for
	getopt which parses long-named-options.

Tue Feb 28 21:12:14 1995  Aubrey Jaffer  (jaffer@jacal)

	* paramlst.scm (parameter-list-expand expanders parms): added.

Mon Feb 27 17:23:54 1995  Aubrey Jaffer  (jaffer@jacal)

	* report.scm (dbutil:print-report): added.

	* comparse.scm (read-command): added.  Reads from a port and
	returns a list of strings: the arguments (and options).

Sat Feb 25 01:05:25 1995  Aubrey Jaffer  (jaffer@jacal)

	* repl.scm (repl:repl): Added loop, conditional on CHAR-READY?
	being PROVIDED?, which reads through trailing white-space.

Sun Feb  5 16:34:03 1995  Aubrey Jaffer  (jaffer@jacal)

	* paramlst.scm ((make-parameter-list parameter-names)): 
	((fill-empty-parameters defaults parameter-list)): 
	((check-parameters checks parameter-list)): 
	((parameter-list->arglist positions arities parameter-list)): 
	((parameter-list-ref parameter-list i)): 
	((adjoin-parameters! parameter-list parameters)): 
	Procedures for making, merging, defaulting, checking and
	converting `parameter lists' (named parameters).
	((getopt->parameter-list argc argv optnames arities aliases)): 
	((getopt->arglist argc argv optnames positions
			 arities defaults checks aliases)): 
	Procedures for converting options and arguments processed by
	getopt to parameter-list or arglist form.

	* dbutil.scm ((make-command-server rdb command-table)): added
	procedure which calls commands and processes parameters.

	* rdms.scm ((make-relational-system base)): add-domain and
	delete-domain commands moved to "dbutil.scm" (create-database).

Fri Feb  3 11:07:46 1995  Aubrey Jaffer  (jaffer@jacal)

	* debug.scm (debug:tracef debug:untracef): removed (duplicates of
	code in "trace.scm").
	(trace-all): utility to trace all defines in a file added.

Thu Jan 19 00:26:14 1995  Aubrey Jaffer  (jaffer@jacal)

	* logical.scm (logbit? logtest): added.

Sun Jan 15 20:38:42 1995  Aubrey Jaffer  (jaffer@jacal)

	* dbutil.scm (dbutil:create-database)): Added parameter
	description tables for "commands".

	* require.scm (software-type): standardize msdos -> ms-dos.

Mon Jan  2 10:26:45 1995  Aubrey Jaffer  (jaffer@jacal)

	* comlist.scm (comlist:atom?): renamed from comlist:atom.

	* scheme48.init	(char->integer integer->char): Now use integers in
	the range 0 to 255.  Fixed several other problems.
	(modulo): Worked around negative modulo bug.

	* Makefile (slib48): `make slib48' loads "scheme48.init", `,dump's
	a scheme48 image file, and creates an `slib48' shell script to
	invoke it.

	* hash.scm (hash:hash-number): no longer does inexact->exact to
	exacts, etc.

	* trnscrpt.scm (read): no longer transcripts eof-objects.

	From: johnm@vlibs.com (John Gerard Malecki)
	* priorque.scm (heap:heapify): internal defines incorrectly
	dependent on order-of-eval replaced with let*.

Thu Dec 22 13:28:16 1994  Aubrey Jaffer  (jaffer@jacal)

	* dbutil.scm (open-database! open-database create-database): This
        enhancement wraps a utility layer on `relational-database' which
        provides:
	  * Automatic loading of the appropriate base-table package when
	    opening a database.
	  * Automatic execution of initialization commands stored in
	    database.
	  * Transparent execution of database commands stored in
	    `*commands*' table in database.

Wed Dec 21 22:53:57 1994  Aubrey Jaffer  (jaffer@jacal)

	* rdms.scm (make-relational-system base): Now more careful about
	protecting read-only databases.

Mon Dec 19 00:06:36 1994  Aubrey Jaffer  (jaffer@jacal)

	* dbutil.scm (dbutil:define-tables): added utility which provides:
	Data definition from Scheme lists for any SLIB
	relational-database.

Sat Dec 17 12:10:02 1994  Aubrey Jaffer  (jaffer@jacal)

	* alistab.scm rdms.scm (make-getter row-eval): evaluation of
	`expression' fields no longer done when retrieved from base
	tables (which made copying of many tables impossible).

	* alistab.scm 
	(write-base): rewrote to not use pretty-print.

	* sc3.scm: removed (only contained last-pair, t, and nil).

	* Template.scm scheme48.init vscm.init (last-pair t nil): added.

Thu Dec  8 00:02:18 1994  Aubrey Jaffer  (jaffer@jacal)

	* mularg.scm pp.scm ratize.scm: copyright line removed from files
	(still lacking terms) less than 12 lines.

	From: johnm@vlibs.com (John Gerard Malecki)
	* sort.scm (sort:sort!): long standing bug in sort! with vector
	argument fixed.

Thu Dec  1 17:10:24 1994  Aubrey Jaffer  (jaffer@jacal)

	* *.scm: Most missing copyright notices supplied.

Sun Nov 27 23:57:41 1994  Aubrey Jaffer  (jaffer@jacal)

	* rdms.scm (make-relational-system base): now checks field types
	when table is opened.  Domains table now has foreign-table field.
	(for-each-row): ordered for-each function added.
	* alistab.scm (ordered-for-each-key supported-key-type?): added.


steering.c
/* steering.c -- UMB Scheme, steering routines

UMB Scheme Interpreter  				$Revision: 3.2 $
Copyright (C) 1988, 1996 William R Campbell

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 1, or (at your option)
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

UMB Scheme was written by Bill Campbell with help from Karl Berry,
Barbara Dixey, Ira Gerstein, Mary Glaser, Kathy Hargreaves, Bill McCabe,
Long Nguyen, Susan Quina, Jeyashree Sivasubram, Bela Sohoni, John Tam 
and Thang Quoc Tran.

For additional information about UMB Scheme, contact the author:

	Bill Campbell
	Department of Mathematics and Computer Science
	University of Massachusetts at Boston
	Harbor Campus
	Boston, MA 02125

	Telephone: 617-287-6449		Internet: bill@cs.umb.edu

*/

#include  <setjmp.h>
#include  <signal.h>

#include "portable.h"
#include "eval.h"
#include "object.h"
#include "primitive.h"
#include "steering.h"
#include "debug.h"
#include "architecture.h"
#include "io.h"
#include "number.h"

/* Public variables. */
Public	jmp_buf	Top_Level_Loop;
Public	jmp_buf	Debugging_Loop;
Public	String	Prompt;

/* Internal routines. */

Private void Initializations();
Private void Steering();
Private	void Load_File();
Private Boolean File_Exists();


#define TOP_LEVEL_PROMPT "\n==> "

#define STANDARD_PRELUDE_PATHNAME "/usr/local/lib/scheme/prelude.scheme"

Private String OPENING  =
"Welcome to UMB Scheme, version      Copyright (c) 1988,1996 William R Campbell.\n\
UMB Scheme comes with ABSOLUTELY NO WARRANTY. This is free software and\n\
you are free to redistribute it under certain conditions.\n\
See the UMB Scheme Release Notes for details. Type Control-d to exit.\n\n";

Private String Rev = "$Revision: 3.2 $";

Private	int	Argc;
Private	char	**Argv;

Public void main( argc , argv )
	int	argc;
	char	*argv[];
{
	Argc = argc;
	Argv = argv;
	Steering();
}

Private	Boolean Init_File_Complete = FALSE;
Public	Boolean Prelude_Started = FALSE;
Public	Boolean	Prelude_Complete = FALSE;

Private void Steering()
{
	Character Opening[400];
	String	Init_Filename = getenv ("SCHEME_INIT");
	Character Dot_Scheme_Filename [256];
	sprintf (Opening, "%s", OPENING);
	sprintf (Dot_Scheme_Filename, "%s/.scheme", getenv ("HOME"));

	Initializations();


	Opening[31] = Rev[11];
	Opening[32] = Rev[12];
	Opening[33] = Rev[13];
	Opening[34] = Rev[14];


	Output( Opening );

	signal( SIGINT  , Handler );
	signal( SIGFPE  , Handler );
	signal( SIGILL  , Handler );
	signal( SIGSEGV , Handler );
	signal( SIGTERM , Handler );

	setjmp( Top_Level_Loop );	/* Return here upon Reset(). */

	clearerr( The_Standard_Input );

	Set_Printing( TRUE );

	Environment_Register = The_Global_Environment;
	State_Register = Nil;
	Value_Register = Nil;
	Expression_Register = Nil;
	Function_Register = Nil;
	Arguments_Register = Nil;
	Reset_Stack( 0 );

	State_Debugged = Nil;
	Value_Debugged = Nil;

	Control_C = FALSE;
	Evaluating = FALSE;
	Evaluation_Broken = FALSE;
	Go_Processed = FALSE;

	if ( ! Prelude_Started )
	{
		Prelude_Started = TRUE;
		Load_File(STANDARD_PRELUDE_PATHNAME);
	}
	Prelude_Complete = TRUE;
	
	if ( ! Init_File_Complete )
	{
		Init_File_Complete = TRUE;

		if (Init_Filename != NULL)
		{
			Load_File (Init_Filename);
		}
		else if (File_Exists (Dot_Scheme_Filename))
		{
			Load_File (Dot_Scheme_Filename);
		}
	}

	while (--Argc > 0)
	{
		Load_File(*++Argv);
	}

	Prompt = TOP_LEVEL_PROMPT;
	Read_Eval_Print( The_Standard_Input );

	if (Arg_Stack_Ptr != 0)
	{
		Panic( "Non-zero argstack pointer on exit" );
		Arg_Stack_Ptr = 0;
	}
}


Private Boolean File_Exists(Filename)

	String      Filename;
{
	FILE * fp;
	
	fp = fopen (Filename, "r");
	if (fp != NULL)
	{
		(void) fclose(fp);
		return( TRUE );
	}
	return( FALSE );
}

Private	void	Load_File(Filename)

	String      Filename;
{
	Make_Constant_String(Filename);
	Push(Value_Register);
	Save(); Load(); Restore();
	Pop(1);
}

Public void Read_Eval_Print( input )
	FILE*	input;	/* C file from which expressions are Read() */
{
	while ( ! Go_Processed )
	{
		if (Get_Printing_State()) Output( Prompt );

		Read( input );

		if (Value_Register == The_Eof_Object) break;

		Push( Value_Register );
		Compile_Object( Top( 1 ));

		Debugger_Activated =  ! Debugging && Debugger_Switched_On;
		Eval( Value_Register, Environment_Register );
		Debugger_Activated = FALSE;

		if (Get_Printing_State()) 
		{
			Output( "\n" );
			(void) Write_Object( Value_Register , 0 );
		}
	}
}


#define ERROR_PREFIX "\nError: "
#define PANIC_PREFIX "\nFatal Error: "


Public void Error( message )
	String	message;
{
	Error_Output( ERROR_PREFIX );
	Error_Output( message );
	Error_Output( ".\n" );
	Break();
}

Public void Error1(message, name)
	String message, name;
{
	Character error_string[256];

	sprintf( error_string, message, name );

	Error_Output( ERROR_PREFIX );
	Error_Output( error_string );
	Error_Output( ".\n" );
	Break();
}

Public void Display_Error(message, object)
	String message;
	Object object;
{
	Error_Output( ERROR_PREFIX );
	Error_Output( message ); 
	Push( Current_Output_Port );
	Current_Output_Port = Current_Error_Port;
	(void) Write_Object( object , 0 );
	Current_Output_Port = Top(1); Pop(1);
	Error_Output( "\n" );
	Break();
}


Public void Panic( message )

	String	message;
{
	Error_Output( PANIC_PREFIX );
	Error_Output( message );
	Error_Output( ".\n" );
	Reset();
}

Public	void Break()
{
	Import	jmp_buf	Eval_Loop;

	if ( Debugger_Activated )
	{
		Debugger_Activated = FALSE;

		if ( Evaluating )
		{
			Evaluation_Broken = TRUE;
			longjmp( Eval_Loop , 1 );
		}
		else
		{
			Reset();
		}
	}
	else if ( Debugging )
	{
		longjmp( Debugging_Loop , 1 );
	}
	else
	{
		Reset();
	}
}


Public	void Reset()
{
	Debugger_Activated = FALSE;
	Error_Output( "\nReset (Use Control-d to quit UMB Scheme)" );
	longjmp( Top_Level_Loop , 1 );
}


Public	void Handler( sig )

	Integer	sig ;
{
	switch ( sig )
	{
	    case SIGINT:
		/* Control-D */

		if ( Allocating )
		{
		   Control_C = TRUE;
		   break;
		}
		else
		{
		   Break();
		}
	
	    case SIGFPE:
		Error( "Floating Point Exception" );

	    case SIGILL:
		Panic( "Illegal Instruction" );

	    case SIGSEGV:
		Panic( "Segmentation Violation" );

	    case SIGTERM:
		Error( "Terminated" );

	    default:
		Panic( "Unhandled Signal" );
	}
}

Private void Initializations()
{
	/* The order of these does matter. */

	Initialize_Architecture();
	Initialize_Object();
	Initialize_Number();
	Initialize_Primitive();
	Initialize_Debug();
}

